msprime/mutations.py

```diff
@@ -56,29 +56,81 @@
-    # that subclasses *must* have an instance variable of the same name.
-    # This is essential for Provenance round-tripping to work.
-    return {
-        key: getattr(self, key)
-        for key in inspect.signature(self.__init__).parameters.keys()
-        if hasattr(self, key)
+def mutation_model_factory(model):
+    """
+    Returns a mutation model corresponding to the specified model.
+    - If model is None, the default mutation model is returned.
+    - If model is a string, return the corresponding model instance.
+    - If model is an instance of MutationModel, return a copy of it.
+    - Otherwise raise a type error.
+    """
+
+    # String names here should match Seq-Gen where possible
+    model_map = {
+        # "slim": SLiMMutationModel(), Needs type argument so can't be string init'd
+        "infinite_alleles": InfiniteAllelesMutationModel(),
+        "binary": BinaryMutationModel(),
+        "jc69": JC69MutationModel(),
+        # "hky": HKYMutationModel(), Needs kappa argument
+        # "f84": F84MutationModel(), Needs kappa argument
+        # "gtr": GTRMutationModel(), Needs relative_rates argument
+        "blosum62": BLOSUM62MutationModel(),
+        "pam": PAMMutationModel(),
     }
 
+    if model is None:
+        model_instance = BinaryMutationModel()
+    elif isinstance(model, str):
+        lower_model = model.lower()
+        if lower_model not in model_map:
+            raise ValueError(
+                "Model '{}' unknown. Choose from {}".format(
+                    model, list(model_map.keys())
+                )
+            )
+        model_instance = model_map[lower_model]
+    elif not isinstance(model, MutationModel):
+        raise TypeError(
+            "Mutation model must be a string or an instance of MutationModel"
+        )
+    else:
+        model_instance = model
+    return model_instance
 
-class MatrixMutationModel(_msprime.MatrixMutationModel):
-    """
-    Superclass of mutation models. Allows you to build your own mutation model.
 
-    TODO document properly.
-    """
+class MutationModel:
+    def asdict(self):
+        # This version of asdict makes sure that we have sufficient parameters
+        # to call the constructor and recreate the class. However, this means
+        # that subclasses *must* have an instance variable of the same name as
+        # each parameter.
+        # This is essential for Provenance round-tripping to work.
+        return {
+            key: getattr(self, key)
+            for key in inspect.signature(self.__init__).parameters.keys()
+            if hasattr(self, key)
+        }
 
-    asdict = safe_asdict
+
+class MatrixMutationModel(_msprime.MatrixMutationModel, MutationModel):
+    """
+    Superclass of the specific matrix mutation models. This class can be used to create
+    a matrix mutation model with custom alleles and probabilities. For details of
+    how this works see :ref:`sec_api_mutation_matrix_models_details`. As this model
```

Comment:
I think this last sentence will be confusing to people reading the documentation for this in isolation. I'm not sure there's much need for it, really, as we can just list the models that can be accepted as strings in the ``model`` argument to mutate. I would delete this sentence and any other ones like it.